core/src/test/java/org/apache/iceberg/TestPartitionSpecUpdate.java

```diff
@@ -0,0 +1,255 @@
+            new String[]{"  1000: data_bucket_removed: void(2)\n", "  1002: id_bucket_removed: void(1)\n"} },
+        new Object[] { 2, new int[]{ 1001, 1000, 1001, 1002, 1002 }, new String[]{"", ""} },
+    };
+  }
+
+  public TestPartitionSpecUpdate(int formatVersion, int[] expectedFieldIds, String[] expectedSpecs) {
+    super(formatVersion);
+    this.expectedFieldIds = expectedFieldIds;
+    this.expectedSpecs = expectedSpecs;
+  }
+
+  @Test
+  public void testCommitUpdatedSpec() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1000, table.spec().lastAssignedFieldId());
+
+    table.updateSpec().clear()
+        .addBucketField("id", 8)
+        .addBucketField("data", 16)
+        .commit();
+
+    Assert.assertEquals("[\n  " +
+        expectedFieldIds[0] + ": id_bucket: bucket[8](1)\n  " +
+        expectedFieldIds[1] + ": data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(expectedFieldIds[2], table.spec().lastAssignedFieldId());
+
+    table.updateSpec().clear()
+        .addTruncateField("data", 8)
+        .commit();
+
+    Assert.assertEquals("[\n  " +
+        expectedFieldIds[3] + ": data_trunc: truncate[8](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(expectedFieldIds[4], table.spec().lastAssignedFieldId());
+  }
+
+  @Test
+  public void testUpdateException() {
+    AssertHelpers.assertThrows(
+        "Should throw IllegalArgumentException if there is an invalid partition field",
+        IllegalArgumentException.class, "Cannot use partition name more than once: id_bucket",
+        () -> table.updateSpec().clear()
+            .addBucketField("id", 8)
+            .addBucketField("id", 16)
+            .commit());
+  }
+
+  @Test
+  public void testAddDuplicateFieldException() {
+    AssertHelpers.assertThrows(
+        "Should throw IllegalArgumentException if adding a duplicate partition field",
+        IllegalArgumentException.class, "Cannot use partition name more than once: data_bucket",
+        () -> table.updateSpec()
+            .addBucketField("data", 16)
+            .commit());
+  }
+
+  @Test
+  public void testAddSamePartitionField() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+
+    if (formatVersion == 1) {
+      table.updateSpec()
+          .addBucketField("data", 16, "data_partition")
+          .commit();
+      Assert.assertEquals("[\n" +
+          "  1000: data_bucket: bucket[16](2)\n" +
+          "  1001: data_partition: bucket[16](2)\n" +
+          "]", table.spec().toString());
+    } else {
+      AssertHelpers.assertThrows(
+          "Should throw IllegalArgumentException if adding a duplicate partition field",
+          IllegalArgumentException.class,
+          "Field Id 1000 has already been used in the existing partition fields",
```

Comment:
I don't see why this would fail with a message about the ID. The field IDs are internal and users should not be exposed to them. The exception in this case should be that the user is trying to add a partition that duplicates another.